grammar/coqpp_parse.mly

```diff
@@ -10,13 +10,59 @@
 
 open Coqpp_ast
 
+let coincide s pat off =
+  let len = String.length pat in
+  let break = ref true in
+  let i = ref 0 in
+  while !break && !i < len do
+    let c = Char.code s.[off + !i] in
+    let d = Char.code pat.[!i] in
+    break := c = d;
+    incr i
+  done;
+  !break
+
+let check_separator sep =
+  if sep <> "" then raise Parsing.Parse_error
+
+let rec parse_user_entry s sep =
+  let l = String.length s in
+  if l > 8 && coincide s "ne_" 0 && coincide s "_list" (l - 5) then
+    let entry = parse_user_entry (String.sub s 3 (l-8)) "" in
+    check_separator sep;
+    Ulist1 entry
+  else if l > 12 && coincide s "ne_" 0 &&
+                   coincide s "_list_sep" (l-9) then
+    let entry = parse_user_entry (String.sub s 3 (l-12)) "" in
+    Ulist1sep (entry, sep)
+  else if l > 5 && coincide s "_list" (l-5) then
+    let entry = parse_user_entry (String.sub s 0 (l-5)) "" in
+    check_separator sep;
+    Ulist0 entry
+  else if l > 9 && coincide s "_list_sep" (l-9) then
+    let entry = parse_user_entry (String.sub s 0 (l-9)) "" in
+    Ulist0sep (entry, sep)
+  else if l > 4 && coincide s "_opt" (l-4) then
+    let entry = parse_user_entry (String.sub s 0 (l-4)) "" in
+    check_separator sep;
+    Uopt entry
+  else if l = 7 && coincide s "tactic" 0 && '5' >= s.[6] && s.[6] >= '0' then
+    let n = Char.code s.[6] - 48 in
+    check_separator sep;
+    Uentryl ("tactic", n)
+  else
+    let s = match s with "hyp" -> "var" | _ -> s in
+    check_separator sep;
+    Uentry s
```

Comment:
Bufffff, this looks like quite an error-prone mess; a table + generic routine should be preferred.

Reply:
That's historical stuff that ought to be removed, and this is just copy-pasting from the corresponding camlp5 file. I know this should be shared, but this would introduce useless dependencies between the camlp5 macros and coqpp. So I went for the easy path, thinking we can get rid of camlp5 soon enough.

Reply:
I can probably seize the opportunity to get rid of this special case though.

Reply:
I also dislike all the hardcoded numbers + indexes; I don't think cleaning this up is a hard requirement but we are introducing very sketchy code in that routine so I'd be good to lessen the technical debt if we can.

Reply:
Ultimately and hopefully all this mess will be replace by Ltac2 that has a sane compositional system for notation scopes.